In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
phone=pd.read_csv(r'C:\Users\adaks\AppData\Local\Programs\Python\Python312\Scripts\Phone Price Prediction\Phone_cleaned_dataset.csv')

In [3]:
phone.head()

Unnamed: 0 Brand me  Ratings  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  \
0           0       LG      4.3  4.0         6.00           48         13   
1           1    IKall      3.4  6.0         4.50           48         12   
2           2    Nokia      4.3  4.0         4.50           64         16   
3           3  Samsung      4.4  6.0         6.40           48         15   
4           4     POCO      4.5  6.0         6.18           35         15   

   Battery_Power    Price  
0           4000  24999.0  
1           4000  15999.0  
2           4000  15000.0  
3           3800  18999.0  
4           3800  18999.0

In [4]:
phone.shape

(521, 9)

In [5]:
phone.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 521 entries, 0 to 520
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     521 non-null    int64  
 1   Brand me       521 non-null    object 
 2   Ratings        521 non-null    float64
 3   RAM            521 non-null    float64
 4   Mobile_Size    521 non-null    float64
 5   Primary_Cam    521 non-null    int64  
 6   Selfi_Cam      521 non-null    int64  
 7   Battery_Power  521 non-null    int64  
 8   Price          521 non-null    float64
dtypes: float64(4), int64(4), object(1)
memory usage: 36.8+ KB


In [6]:
backup=phone.copy()

Converting brand name into numerical data using label encoding

In [7]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()

In [8]:
phone['Brand me']=label_encoder.fit_transform(phone['Brand me'])

In [9]:
phone.head()

Unnamed: 0  Brand me  Ratings  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  \
0           0        26      4.3  4.0         6.00           48         13   
1           1        17      3.4  6.0         4.50           48         12   
2           2        39      4.3  4.0         4.50           64         16   
3           3        48      4.4  6.0         6.40           48         15   
4           4        42      4.5  6.0         6.18           35         15   

   Battery_Power    Price  
0           4000  24999.0  
1           4000  15999.0  
2           4000  15000.0  
3           3800  18999.0  
4           3800  18999.0

In [10]:
phone.drop('Unnamed: 0',axis=1)

Brand me  Ratings  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  \
0          26      4.3  4.0         6.00           48         13   
1          17      3.4  6.0         4.50           48         12   
2          39      4.3  4.0         4.50           64         16   
3          48      4.4  6.0         6.40           48         15   
4          42      4.5  6.0         6.18           35         15   
..        ...      ...  ...          ...          ...        ...   
516        48      4.6  4.0         6.20           35          8   
517         4      4.4  6.0         6.39           48         12   
518        24      3.8  6.0         4.54           48         12   
519         0      4.4  3.0         6.20           48          1   
520         2      3.5  6.0         4.50           64         15   

     Battery_Power    Price  
0             4000  24999.0  
1             4000  15999.0  
2             4000  15000.0  
3             3800  18999.0  
4             3800  18999.0  
..             ...      ...  
516           3500  53990.0  
517           3800  31999.0  
518           2800   1299.0  
519           3800   9790.0  
520           1050    799.0  

[521 rows x 8 columns]

In [12]:
phone=phone.drop('Ratings',axis=1)

In [13]:
phone.head()

Unnamed: 0  Brand me  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  \
0           0        26  4.0         6.00           48         13   
1           1        17  6.0         4.50           48         12   
2           2        39  4.0         4.50           64         16   
3           3        48  6.0         6.40           48         15   
4           4        42  6.0         6.18           35         15   

   Battery_Power    Price  
0           4000  24999.0  
1           4000  15999.0  
2           4000  15000.0  
3           3800  18999.0  
4           3800  18999.0

In [14]:
phone=phone.drop('Unnamed: 0',axis=1)

In [15]:
phone.head()

Brand me  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  Battery_Power    Price
0        26  4.0         6.00           48         13           4000  24999.0
1        17  6.0         4.50           48         12           4000  15999.0
2        39  4.0         4.50           64         16           4000  15000.0
3        48  6.0         6.40           48         15           3800  18999.0
4        42  6.0         6.18           35         15           3800  18999.0

Splitting Data into input features and output

In [17]:
X = phone[['Brand me', 'RAM','Mobile_Size', 'Primary_Cam', 'Selfi_Cam', 'Battery_Power']]
y = phone['Price']

In [18]:
X

Brand me  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  Battery_Power
0          26  4.0         6.00           48         13           4000
1          17  6.0         4.50           48         12           4000
2          39  4.0         4.50           64         16           4000
3          48  6.0         6.40           48         15           3800
4          42  6.0         6.18           35         15           3800
..        ...  ...          ...          ...        ...            ...
516        48  4.0         6.20           35          8           3500
517         4  6.0         6.39           48         12           3800
518        24  6.0         4.54           48         12           2800
519         0  3.0         6.20           48          1           3800
520         2  6.0         4.50           64         15           1050

[521 rows x 6 columns]

In [19]:
y

0      24999.0
1      15999.0
2      15000.0
3      18999.0
4      18999.0
        ...   
516    53990.0
517    31999.0
518     1299.0
519     9790.0
520      799.0
Name: Price, Length: 521, dtype: float64

Splitting data into training and testing data

In [20]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Training the Decision Tree regression model

In [22]:
model=DecisionTreeRegressor()
model.fit(X_train,y_train)

DecisionTreeRegressor()

Our model has been trained, now checking the accuracy

In [23]:
y_pred=model.predict(X_test)

In [24]:
mse=mean_squared_error(y_test,y_pred)
print(f'Mean Squared Error:{mse}')

Mean Squared Error:332452995.27261907


In [25]:
from sklearn.metrics import mean_absolute_error, r2_score

In [26]:
r2=r2_score(y_test,y_pred)

In [27]:
print(f'R2 Score: {r2}')

R2 Score: 0.9101717808595995


We will consider the random state that has maximum accuracy

In [31]:
score=[]
for i in range(100):
    X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=i)
    model=DecisionTreeRegressor()
    model.fit(X_train,y_train)
    y_pred=model.predict(X_test)
    #print(r2_score(y_test,y_pred),i)
    score.append(r2_score(y_test,y_pred))

In [33]:
np.argmax(score)

4

In [34]:
score[np.argmax(score)]

0.9957208407376105

We get a maximum accuracy of 0.9957 at randome state=4